pkg/master/controller.go

```diff
@@ -0,0 +1,218 @@
+// namespace, and provide the IP repair check on service PortalIPs
+type Controller struct {
+	NamespaceRegistry namespace.Registry
+	ServiceRegistry   service.Registry
+	ServiceIPRegistry service.IPRegistry
+	EndpointRegistry  endpoint.Registry
+	PortalNet         *net.IPNet
+
+	PortalIPInterval time.Duration
+	EndpointInterval time.Duration
+
+	PublicIP net.IP
+
+	ServiceIP         net.IP
+	ServicePort       int
+	PublicServicePort int
+
+	ReadOnlyServiceIP         net.IP
+	ReadOnlyServicePort       int
+	PublicReadOnlyServicePort int
+
+	runner *util.Runner
+}
+
+// Start begins the core controller loops that must exist for bootstrapping
+// a cluster.
+func (c *Controller) Start() {
+	if c.runner != nil {
+		return
+	}
+
+	repair := servicecontroller.NewRepair(c.PortalIPInterval, c.ServiceRegistry, c.PortalNet, c.ServiceIPRegistry)
+
+	// run all of the controllers once prior to returning from Start.
+	if err := repair.RunOnce(); err != nil {
+		glog.Errorf("Unable to perform initial IP allocation check: %v", err)
+	}
+	if err := c.UpdateKubernetesService(); err != nil {
+		glog.Errorf("Unable to perform initial Kubernetes service initialization: %v", err)
+	}
+	if err := c.UpdateKubernetesROService(); err != nil {
+		glog.Errorf("Unable to perform initial Kubernetes RO service initialization: %v", err)
+	}
+
+	c.runner = util.NewRunner(c.RunKubernetesService, c.RunKubernetesROService, repair.RunUntil)
+	c.runner.Start()
+}
+
+// RunKubernetesService periodically updates the kubernetes service
+func (c *Controller) RunKubernetesService(ch chan struct{}) {
+	util.Until(func() {
+		if err := c.UpdateKubernetesService(); err != nil {
+			util.HandleError(fmt.Errorf("unable to sync kubernetes service: %v", err))
+		}
+	}, c.EndpointInterval, ch)
+}
+
+// UpdateKubernetesService attempts to update the default Kube service.
+func (c *Controller) UpdateKubernetesService() error {
+	// Update service & endpoint records.
+	// TODO: when it becomes possible to change this stuff,
+	// stop polling and start watching.
+	// TODO: add endpoints of all replicas, not just the elected master.
+	if err := c.CreateNamespaceIfNeeded(api.NamespaceDefault); err != nil {
+		return err
+	}
+	if c.ServiceIP != nil {
+		if err := c.CreateMasterServiceIfNeeded("kubernetes", c.ServiceIP, c.ServicePort); err != nil {
+			return err
+		}
+		if err := c.SetEndpoints("kubernetes", c.PublicIP, c.PublicServicePort); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// RunKubernetesROService periodically updates the kubernetes RO service
+func (c *Controller) RunKubernetesROService(ch chan struct{}) {
+	util.Until(func() {
+		if err := c.UpdateKubernetesROService(); err != nil {
+			util.HandleError(fmt.Errorf("unable to sync kubernetes RO service: %v", err))
+		}
+	}, c.EndpointInterval, ch)
+}
+
+// UpdateKubernetesROService attempts to update the default Kube read-only service.
+func (c *Controller) UpdateKubernetesROService() error {
```

Comment:
It could be - I mostly renamed existing code.

----- Original Message -----
> > +		}
> > +	}
> > +	return nil
> > +}
> > +
> > +// RunKubernetesROService periodically updates the kubernetes RO service
> > +func (c *Controller) RunKubernetesROService(ch chan struct{}) {
> > +	util.Until(func() {
> > +		if err := c.UpdateKubernetesROService(); err != nil {
> > +			util.HandleError(fmt.Errorf("unable to sync kubernetes RO service: %v",
> > err))
> > +		}
> > +	}, c.EndpointInterval, ch)
> > +}
> > +
> > +// UpdateKubernetesROService attempts to update the default Kube read-only
> > service.
> > +func (c *Controller) UpdateKubernetesROService() error {
> 
> This appears to be structurally identical to UpdateKubernetesService().  Any
> reason not to make one parameterized function?
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/7704/files#r29736534